src/packaging/wheelfile.py

```diff
@@ -0,0 +1,544 @@
+        self.hash_algorithm = hash_algorithm
+        self._compress_type = ZIP_DEFLATED if compress else ZIP_STORED
+
+        if metadata:
+            self.metadata = metadata
+        elif isinstance(path_or_fd, (str, PathLike)):
+            filename = Path(path_or_fd).name
+            self.metadata = WheelMetadata.from_filename(filename)
+        else:
+            raise WheelError("path_or_fd is not a path, and metadata was not provided")
+
+        if hash_algorithm not in hashlib.algorithms_available:
+            raise ValueError(f"Hash algorithm {hash_algorithm!r} is not available")
+        elif hash_algorithm in ("md5", "sha1"):
+            raise ValueError(
+                f"Weak hash algorithm ({hash_algorithm}) is not permitted by PEP 427"
+            )
+
+        self._dist_info_dir = f"{self.metadata.name}-{self.metadata.version}.dist-info"
+        self._data_dir = f"{self.metadata.name}-{self.metadata.version}.data"
+        self._record_path = f"{self._dist_info_dir}/RECORD"
+        self._record_entries: dict[str, WheelRecordEntry] = OrderedDict()
+
+    def __enter__(self) -> WheelWriter:
+        self._zip = ZipFile(self.path_or_fd, "w", compression=self._compress_type)
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException],
+        exc_val: BaseException,
+        exc_tb: TracebackType,
+    ) -> None:
+        try:
+            if not exc_type:
+                if f"{self._dist_info_dir}/WHEEL" not in self._record_entries:
+                    self._write_wheelfile()
+
+                self._write_record()
+        finally:
+            self._zip.close()
+
+    def _write_record(self) -> None:
+        data = StringIO()
+        writer = csv.writer(data, delimiter=",", quotechar='"', lineterminator="\n")
+        writer.writerows(
+            [
+                (
+                    fname,
+                    entry.hash_algorithm + "=" + _encode_hash_value(entry.hash_value),
+                    entry.filesize,
+                )
+                for fname, entry in self._record_entries.items()
+            ]
+        )
+        writer.writerow((self._record_path, "", ""))
+        self.write_distinfo_file("RECORD", data.getvalue())
+
+    def _write_wheelfile(self) -> None:
+        buffer = BytesIO()
+        write_wheelfile(
+            buffer,
+            generator=self.generator,
+            metadata=self.metadata,
+            root_is_purelib=self.root_is_purelib,
+        )
+        self.write_distinfo_file("WHEEL", buffer.getvalue())
+
+    def write_metadata(self, items: Iterable[tuple[str, str]]) -> None:
+        msg = Message(policy=_email_policy)
+        for key, value in items:
+            key = key.title()
+            if key == "Description":
+                msg.set_payload(value, "utf-8")
+            else:
+                msg.add_header(key, value)
+
+        if "Metadata-Version" not in msg:
+            msg["Metadata-Version"] = "2.1"
```

Comment:
Why default to 2.1? 2.3 is the current version, and in general, a default can't be safely updated, so IMO this should be required.

Reply:
I wasn't aware; I just copied this from the old code.